### PR TITLE
Switch to Cuckoo Filter

### PR DESCRIPTION
**Pull request by kazuho**
Simply swaps GCS with Cuckoo Filter.

The difference from #413 (which this PR is based upon) is that this PR retains the possibility of the client sending a digest of fresh resources without ETag values being included as part of the hash.

This was considered necessary so that an endpoint can determine if it needs to push a resource without the knowledge of the ETag value of the resource. For example, by URL being the only source of the hash function, a H2 proxy that just holds a list of preload URLs can determine which ones in the list needs to be pushed, _and then_ fetch them from origin servers to push them.

### CODE REVIEW

**Comment by kazuho** [comment]
@yoavweiss @mnot PTAL

File: draft-ietf-httpbis-cache-digest.md [hunk 1]
```diff
@@ -375,10 +487,7 @@ The header may represent the cache state of a client or that of a proxy, dependi
 
 # Acknowledgements
 
-Thanks to Adam Langley and Giovanni Bajo for their explorations of Golomb-coded sets. In
-particular, see
-<http://giovanni.bajo.it/post/47119962313/golomb-coded-sets-smaller-than-bloom-filters>, which
-refers to sample code.
+Thanks to Joav Weiss for his idea and text to use Cuckoo Filter.
```
Comment (yoavweiss) [inline draft-ietf-httpbis-cache-digest.md]: s/Joav/Yoav/ :D

Comment (kazuho) [reply draft-ietf-httpbis-cache-digest.md]: I'm terribly sorry for the mistake. Fixed in f8e63e2.

**Comment by kazuho** [comment]
@yoavweiss Thank you for checking!

File: draft-ietf-httpbis-cache-digest.md [hunk 2]
```diff
@@ -182,64 +170,171 @@ response was fresh or stale.
...
+        3. Add `f` to `position_start`.
+    4. Let `e` be a random number from 0 to `b`.
+    5. Substract `f` * (`b` - `e`) from `position_start`.
+    6. Let `bits` be `f` bits from `digest_value` starting at `position_start`.
+    7. Let `fingerprint` be the value of bits, read as big endian.
+    8. Set `bits` to `dest_fingerprint`.
+    9. Set `dest_fingerprint` to `fingerprint`.
+    10. Let `h` be {{hash2}} with `h`, `dest_fingerprint` and `N` as inputs.
+    11. Substract 1 from `maxcount`.
+13. Substract `f` from `position_start`.
```
Comment (mnot) [inline draft-ietf-httpbis-cache-digest.md]: Double check that the generated text and HTML are correct. In the past, list numbers higher than 9 needed to wrap around to 0 to work correctly; not sure if that's been fixed or not.

**Comment by kazuho** [comment]
My bad. I have accidentally removed the branch that caused the PR to get closed.

File: draft-ietf-httpbis-cache-digest.md (hunk 2 above)
Comment (kazuho) [reply draft-ietf-httpbis-cache-digest.md]: Thank you for pointing that out.

The numbers were correctly handled by the parser, however the numbers in the markdown were wrong.

The issue has been fixed in 8469d48.

**Comment by kazuho** [comment]
Since I accidentally removed the branch, it has become impossible to continue reviewing the changes in this branch. I apologize for the inconvenience.

Fortunately, @yoavweiss and @mnot have already given their comments on the PR. Considering that, I have merged the PR to master in 63284be (FWIW, you can click the commit hash to see the changes that went into master).

If anybody sees an issue, I would appreciate if they could open a new issue (or a PR).

**Comment by martinthomson** [comment]
Nothing major, but s/Substract/Subtract/ig

**Comment by kazuho** [comment]
@martinthomson Thank you for pointing that out. I was able to fix that in a8f1ca0 prior to submitting -03.

**Pull request closed without merging**